File: medium/86. Partition List/86. Partition List.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function partition(head: ListNode | null, x: number): ListNode | null {
    const low = new ListNode(0)
    let lowTail = low

    const great = new ListNode(0)
    let greatTail = great

    while (head) {
        if (head.val < x) {
            lowTail.next = new ListNode(head.val)
            lowTail = lowTail.next
        } else {
            greatTail.next = new ListNode(head.val)
            greatTail = greatTail.next
        }

        head = head.next
    }

    lowTail.next = great.next
    greatTail.next = null

    return low.next
};

export { }

File: medium/86. Partition List/86. Partition List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        lowest_nodes = ListNode(0)
        go_lowest_nodes = lowest_nodes

        greatest_nodes = ListNode(0)
        go_greatest_nodes = greatest_nodes

        while head:
            val = ListNode(head.val)
            if head.val < x:
                go_lowest_nodes.next = val
                go_lowest_nodes = go_lowest_nodes.next
            else:
                go_greatest_nodes.next = val
                go_greatest_nodes = go_greatest_nodes.next

            head = head.next

        go_lowest_nodes.next = greatest_nodes.next
        return lowest_nodes.next
